File: core/src/main/java/com/canyon/commons/SerializeUtils.java
package com.canyon.commons;

import java.beans.XMLDecoder;
import java.beans.XMLEncoder;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

public class SerializeUtils {
	public static void serializeToFile(String path, Object obj) {
		if (obj == null)
			return;

		String dir = IOUtils.getDirectory(path);
		IOUtils.createDirectory(dir);

		ObjectOutputStream out = null;
		try {
			out = new ObjectOutputStream(new FileOutputStream(path));
			out.writeObject(obj);
			out.flush();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			if (out != null) {
				try {
					out.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		}
	}

	@SuppressWarnings({ "unchecked" })
	public static <T> T deserializeFromFile(String path) {
		T result = null;
		if (!IOUtils.isFile(path))
			return result;

		ObjectInputStream in = null;
		try {
			in = new ObjectInputStream(new FileInputStream(path));
			result = (T) in.readObject();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			if (in != null) {
				try {
					in.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		}
		return result;
	}

	public static byte[] serializeToBytes(Object obj) {
		byte[] result = null;
		if (obj == null)
			return result;

		ByteArrayOutputStream stream = null;
		ObjectOutputStream out = null;
		try {
			stream = new ByteArrayOutputStream();
			out = new ObjectOutputStream(stream);
			out.writeObject(obj);
			result = stream.toByteArray();
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			if (out != null) {
				try {
					out.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			if (stream != null) {
				try {
					stream.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		}
		return result;
	}

	@SuppressWarnings("unchecked")
	public static <T> T deserializeFromBytes(byte[] data) {
		T result = null;
		if (data == null || data.length == 0)
			return result;

		ObjectInputStream in = null;
		try {
			in = new ObjectInputStream(new ByteArrayInputStream(data));
			result = (T) in.readObject();
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		} finally {
			if (in != null) {
				try {
					in.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		}

		return result;
	}

	public static void xmlSerializeToFile(String path, Object obj) {
		if (obj == null)
			return;

		String dir = IOUtils.getDirectory(path);
		IOUtils.createDirectory(dir);

		XMLEncoder encoder = null;
		try {
			encoder = new XMLEncoder(new FileOutputStream(path));
			encoder.writeObject(obj);
			encoder.flush();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} finally {
			if (encoder != null) {
				encoder.close();
			}
		}
	}

	@SuppressWarnings("unchecked")
	public static <T> T xmlDeserializeFromFile(String path) {
		T result = null;
		if (!IOUtils.isFile(path))
			return result;

		XMLDecoder decoder = null;
		try {
			decoder = new XMLDecoder(new FileInputStream(path));
			result = (T) decoder.readObject();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} finally {
			if (decoder != null) {
				decoder.close();
			}
		}
		return result;
	}
}

File: commons/src/main/kotlin/com/canyon/commons/SerializeUtils.kt
package com.canyon.commons

import java.beans.XMLDecoder
import java.beans.XMLEncoder
import java.io.*

object SerializeUtils {
    fun serializeToFile(path: String, obj: Any) {
        val dir = IOUtils.getDirectory(path)
        IOUtils.createDirectory(dir)
        FileOutputStream(path).use { fos->
            ObjectOutputStream(fos).use { out->
                out.writeObject(obj)
                out.flush()
            }
        }
    }

    fun <T> deserializeFromFile(path: String): T? {
        var result: T? = null
        if (!IOUtils.isFile(path))
            return result

        FileInputStream(path).use { fis->
            ObjectInputStream(fis).use { ois->
                result = ois.readObject() as T
            }
        }
        return result
    }

    fun serializeToBytes(obj: Any?): ByteArray? {
        var result: ByteArray? = null
        if (obj == null)
            return result

        var stream: ByteArrayOutputStream? = null
        var out: ObjectOutputStream? = null
        try {
            stream = ByteArrayOutputStream()
            out = ObjectOutputStream(stream)
            out.writeObject(obj)
            result = stream.toByteArray()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            if (out != null) {
                try {
                    out.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
            if (stream != null) {
                try {
                    stream.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }
        return result
    }

    fun <T> deserializeFromBytes(data: ByteArray?): T? {
        var result: T? = null
        if (data == null || data.size == 0)
            return result

        var `in`: ObjectInputStream? = null
        try {
            `in` = ObjectInputStream(ByteArrayInputStream(data))
            result = `in`.readObject() as T
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            if (`in` != null) {
                try {
                    `in`.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }

        return result
    }

    fun xmlSerializeToFile(path: String, obj: Any?) {
        if (obj == null)
            return

        val dir = IOUtils.getDirectory(path)
        IOUtils.createDirectory(dir)

        var encoder: XMLEncoder? = null
        try {
            encoder = XMLEncoder(FileOutputStream(path))
            encoder.writeObject(obj)
            encoder.flush()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } finally {
            encoder?.close()
        }
    }

    fun <T> xmlDeserializeFromFile(path: String): T? {
        var result: T? = null
        if (!IOUtils.isFile(path))
            return result

        var decoder: XMLDecoder? = null
        try {
            decoder = XMLDecoder(FileInputStream(path))
            result = decoder.readObject() as T
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } finally {
            decoder?.close()
        }
        return result
    }
}
